### ISSUE TITLE
Unflushed stderr doesn't get captured via CliRunner

### ISSUE THREAD

**Issue by ericfrederich** [pallets/click#2647]
CliRunner flushes stdout but not stderr.

To reproduce:
Invoke a command that uses `print` with `file=sys.stderr` which doesn't flush by default (`click.secho` flushes by default).

Expected results:
Anything written to stderr should be captured in the result of a `CliRunner.invoke()`

Environment:

- Python version: 3.11
- Click version: 8.1.7 as well as `main` branch here on GitHub

**Comment by ericfrederich** [comment]
Fyi... this was the code I used to discover the bug.  On the pull request I ended up just modifying an existing test.

```python
import click
import pytest
from click.testing import CliRunner


@click.command()
@click.option("--out/--no-out")
@click.option("--err/--no-err")
def dummy_cmd(out: bool, err: bool):
    import sys

    print("this is stdout", file=sys.stdout)
    print("this is stderr", file=sys.stderr)
    if out:
        print("Extra out!", file=sys.stdout)
    if err:
        print("Extra err!", file=sys.stderr)


@pytest.mark.parametrize("out_option", [None, True, False])
@pytest.mark.parametrize("err_option", [None, True, False])
def test_dummy_cli(out_option: str | None, err_option: str | None):
    runner = CliRunner(mix_stderr=False)
    args = []
    expected_out_lines = 1
    expected_err_lines = 1
    if out_option is not None:
        if out_option:
            args.append("--out")
            expected_out_lines += 1
        else:
            args.append("--no-out")
    if err_option is not None:
        if err_option:
            args.append("--err")
            expected_err_lines += 1
        else:
            args.append("--no-err")
    result = runner.invoke(dummy_cmd, [*args])

    if result.exception:
        raise result.exception

    if result.exit_code != 0:
        raise AssertionError(f"Non-zero return code: {result.exit_code!r}\noutput:\n{result.output}")

    assert len(result.stdout.splitlines()) == expected_out_lines
    assert len(result.stderr.splitlines()) == expected_err_lines
```

**Comment by davidism** [comment]
Duplicate of #2636. If you're using things that don't flush, like `print`, then they don't flush. Hiding that by doing it in the test runner just hides the potentially incorrect output behavior you're using.

**Comment by davidism** [comment]
You're welcome to use print to street, you just have to remember to flush in that case. That's true regardless of if you're using click.

**Comment by davidism** [comment]
Another way to put it is: you've need to explain why your suggested behavior is _correct_. It's certainly _different than current_, but that doesn't mean it's _more correct_.

What I know is that stdout is flushed, and stderr isn't. I don't know why that's the case, but I do know that this is the behavior I've encountered all over the place, and if I want stderr to show up at a specific time I need to flush manually. This is what `click.echo` does automatically. If what I know is incorrect, then you need to show me authoritative sources that say so. Otherwise, as I said, it would just mean we're hiding the actual behavior of the calls you're making.

**Comment by ericfrederich** [comment]
Thanks, I think understand what you're saying.  Are you saying the CliRunner flushed stdout to mimic the behavior you've encountered all over the place?

If click.echo forces a flush every time though, then why does the CliRunner need to flush as well?  Why do the tests fail when that flush in the CliRunner is removed?

**Comment by kdeldycke** [comment]
This has been fixed in v8.2.1 by https://github.com/pallets/click/pull/2933.
